cpmpy/transformations/linearize.py

```diff
@@ -0,0 +1,411 @@
+        raise Exception("Numeric constants or numeric variables not allowed as base constraint")
+
+    if cpm_expr.name == "or":
+        if all(arg.is_bool() for arg in cpm_expr.args):
+            return [sum(cpm_expr.args) >= 1]
+        raise Exception("Numeric constants or numeric variables not allowed as base constraint")
+
+    if cpm_expr.name == "xor":
+        assert len(cpm_expr.args) == 2, "Only supports xor with 2 vars"
+        if all(arg.is_bool() for arg in cpm_expr.args):
+            return [sum(cpm_expr.args) == 1]
+        raise Exception("Numeric constants or numeric variables not allowed as base constraint")
+
+    if cpm_expr.name == "->":
+        cond, sub_expr = cpm_expr.args
+        var_cons = []
+        if not cond.is_bool() or not sub_expr.is_bool():
+            raise Exception(
+                f"Numeric constants or numeric variables not allowed as base constraint, cannot linearize {cpm_expr}")
+        lin_comps = linearize_constraint(sub_expr)
+        lin_exprs = []
+        for l_expr in lin_comps:
+            lhs, rhs = l_expr.args
+            if lhs.name == "mul" and _wsum_should(lhs):
+                lhs = Operator("wsum", _wsum_make(lhs))
+
+            if isinstance(rhs, _NumVarImpl):
+                # Vars on rhs of linexpr not supported in indicator constraints
+                if isinstance(lhs, _NumVarImpl):
+                    new_lhs = lhs + (-1) * rhs
+                elif lhs.name == "sum":
+                    new_lhs = Operator("wsum", [[1]*len(lhs.args) + [-1], lhs.args + [rhs]])
+                elif lhs.name == "wsum":
+                    new_lhs = lhs + (-1 * rhs)
+                else:
+                    raise Exception(f"Unsupported expression {lhs} on right hand side of implication {l_expr}, resulting from linearization of {cpm_expr}")
+                lin_exprs += [Comparison(l_expr.name, new_lhs, 0)]
+            else:
+                lin_exprs += [l_expr]
+
+        return var_cons + [cond.implies(l_expr) for l_expr in lin_exprs]
+
+    # Binary operators
+    if cpm_expr.name == "<":
+        # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/issues/97
+        lhs, rhs = cpm_expr.args
+        if lhs.name == "wsum":
+            return [Operator("wsum", [lhs.args[0] + [1], lhs.args[1] + [1]]) <= rhs]
+        else:
+            cons = lhs + 1 <= rhs
+        cons = flatten_constraint(cons)
+        cons = linearize_constraint(cons)
+        return cons
+
+    if cpm_expr.name == ">":
+        # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/issues/97
+        lhs, rhs = cpm_expr.args
+        if lhs.name == "wsum":
+            return [Operator("wsum", [lhs.args[0] + [-1], lhs.args[1] + [1]]) >= rhs]
+        else:
+            cons = lhs - 1 >= rhs
+        cons = flatten_constraint(cons)
+        cons = linearize_constraint(cons)
+        return cons
+
+    if cpm_expr.name == "!=":
+        lhs, rhs = cpm_expr.args
```

Comment:
add smth like
-- TODO: this can be optimized, example cases:
  - if rhs is constant and lhs is var, then make a Boolean for every value of lhs, add sum(bool_lhs) = 1, also sum(bool_lhs*idx_lhs)=lhs potentially, and then that bool_lhs[rhs] <= 0.  While probably not very understandable, this is how it seems to be implemented in numberjacks mipwrapper: https://github.com/eomahony/Numberjack/blob/53fa9e994a36f881ffd320d8d04158097190aad8/Numberjack/solvers/MipWrapper/MipWrapper.cpp#L95
  - other case of rhs is constant but lhs is expression, make `lhs == aux` for an appropriate auxiliary variable then do the above

(I'm not suggesting to implement this necessarily, just to add the above for future reference...)